codex/logutils.nim

```diff
@@ -0,0 +1,211 @@
+## so we could reduce this even further by not specifying a `LogFormat`:
+## ```nim
+## type
+##   BlockAddress* = object
+##     case leaf*: bool
+##     of true:
+##       treeCid* {.serialize.}: Cid
+##       index* {.serialize.}: Natural
+##     else:
+##       cid* {.serialize.}: Cid
+##
+## logutils.formatIt(BlockAddress): %it
+##
+## # chronicles textlines output
+## TRC test tid=14400673 ba="{\"treeCid\":\"zb2rhgsDE16rLtbwTFeNKbdSobtKiWdjJPvKEuPgrQAfndjU1\",\"index\":0}"
+## # chronicles json output
+## {"lvl":"TRC","msg":"test","tid":14400673,"ba":{"treeCid":"zb2rhgsDE16rLtbwTFeNKbdSobtKiWdjJPvKEuPgrQAfndjU1","index":0}}
+## ```
+
+import std/options
+import std/sequtils
+import std/strutils
+import std/sugar
+import std/typetraits
+
+import pkg/chronicles except toJson, `%`
+from pkg/libp2p import Cid, MultiAddress, `$`
+import pkg/questionable
+import pkg/stew/byteutils
+import pkg/stint
+import pkg/upraises
+
+import ./utils/json
+
+export byteutils
+export chronicles except toJson, formatIt, `%`
+export questionable
+export sequtils
+export strutils
+export sugar
+export upraises
+export json
+
+func shortLog*(long: string, ellipses = "*", start = 3, stop = 6): string =
+  ## Returns compact string representation of ``long``.
+  var short = long
+  let minLen = start + ellipses.len + stop
+  if len(short) > minLen:
+    short.insert(ellipses, start)
+
+    when (NimMajor, NimMinor) > (1, 4):
+      short.delete(start + ellipses.len .. short.high - stop)
+    else:
+      short.delete(start + ellipses.len, short.high - stop)
+
+  short
+
+func shortHexLog*(long: string): string =
+  if long[0..1] == "0x": result &= "0x"
+  result &= long[2..long.high].shortLog("..", 4, 4)
+
+func short0xHexLog*[N: static[int], T: array[N, byte]](v: T): string =
+  v.to0xHex.shortHexLog
+
+func short0xHexLog*[T: distinct](v: T): string =
+  type BaseType = T.distinctBase
+  BaseType(v).short0xHexLog
+
+func short0xHexLog*[U: distinct, T: seq[U]](v: T): string =
+  type BaseType = U.distinctBase
+  "@[" & v.map(x => BaseType(x).short0xHexLog).join(",") & "]"
+
+func to0xHexLog*[T: distinct](v: T): string =
+  type BaseType = T.distinctBase
+  BaseType(v).to0xHex
+
+func to0xHexLog*[U: distinct, T: seq[U]](v: T): string =
+  type BaseType = U.distinctBase
+  "@[" & v.map(x => BaseType(x).to0xHex).join(",") & "]"
+
+proc formatTextLineSeq*(val: seq[string]): string =
+  "@[" & val.join(", ") & "]"
+
+template formatIt*(format: LogFormat, T: typedesc, body: untyped) {.dirty.} =
+  # Provides formatters for logging with Chronicles for the given type and
+  # `LogFormat`.
+  # NOTE: `seq[T]`, `Option[T]`, and `seq[Option[T]]` are overriddden
+  # since the base `setProperty` is generic using `auto` and conflicts with
+  # providing a generic `seq` and `Option` override.
+  when format == LogFormat.json:
+    proc formatJsonOption(val: ?T): JsonNode =
+      if it =? val:
+        json.`%`(body)
+      else:
+        newJNull()
+
```

Comment:
Oddly enough, if I add:

```nim
import questionable/results

proc setProperty*(r: var JsonRecord, key: string, opt: ?!T) =
  # ...
```
The compiler complains that it cannot resolve the symbol `?!`. The same happens if I use the long hand `Result[T, ref CatchableError]` as well. `nimsuggest` offers a further explanation that this may be due to a circular depdency, BUT HOW???

```shell
      undeclared identifier: '?!'
candidates (edit distance, scope distance); see '--spellSuggest': 
 (1, 3): '!' [template declared in /Users/egonat/repos/status-im/nim-codex/vendor/questionable/questionable/options.nim(23, 10)]
 (1, 3): '?' [template declared in /Users/egonat/repos/status-im/nim-codex/vendor/questionable/questionable/options.nim(17, 10)]
This might be caused by a recursive module dependency:
/Users/egonat/repos/status-im/nim-codex/codex/logutils.nim imports /Users/egonat/repos/status-im/nim-codex/codex/logutils.nim
/Users/egonat/repos/status-im/nim-codex/codex/logutils.nim imports /Users/egonat/repos/status-im/nim-codex/codex/codex.nim
/Users/egonat/repos/status-im/nim-codex/codex/codex.nim imports /Users/egonat/repos/status-im/nim-codex/codex/logutils.nim
/Users/egonat/repos/status-im/nim-codex/codex/logutils.nim imports /Users/egonat/repos/status-im/nim-codex/codex/codex.nim
/Users/egonat/repos/status-im/nim-codex/codex/codex.nim imports /Users/egonat/repos/status-im/nim-codex/codex/codex.nim
/Users/egonat/repos/status-im/nim-codex/codex/codex.nim imports /Users/egonat/repos/status-im/nim-codex/codex/logutils.nim
/Users/egonat/repos/status-im/nim-codex/codex/logutils.nim imports /Users/egonat/repos/status-im/nim-codex/codex/node.nim
/Users/egonat/repos/status-im/nim-codex/codex/node.nim imports /Users/egonat/repos/status-im/nim-codex/codex/node.nim
/Users/egonat/repos/status-im/nim-codex/codex/node.nim imports /Users/egonat/repos/status-im/nim-codex/codex/node.nim
/Users/egonat/repos/status-im/nim-codex/codex/node.nim imports /Users/egonat/repos/status-im/nim-codex/codex/node.nim
/Users/egonat/repos/status-im/nim-codex/codex/node.nim imports /Users/egonat/repos/status-im/nim-codex/codex/node.nim
/Users/egonat/repos/status-im/nim-codex/codex/node.nim imports /Users/egonat/repos/status-im/nim-codex/codex/node.nim
/Users/egonat/repos/status-im/nim-codex/codex/node.nim imports /Users/egonat/repos/status-im/nim-codex/codex/codex.nim
/Users/egonat/repos/status-im/nim-codex/codex/codex.nim imports /Users/egonat/repos/status-im/nim-codex/codex/logutils.nim
/Users/egonat/repos/status-im/nim-codex/codex/logutils.nim imports /Users/egonat/repos/status-im/nim-codex/codex/logutils.nim
/Users/egonat/repos/status-im/nim-codex/codex/logutils.nim imports /Users/egonat/repos/status-im/nim-codex/codex/codex.nim
/Users/egonat/repos/status-im/nim-codex/codex/codex.nim imports /Users/egonat/repos/status-im/nim-codex/codex/logutils.nim
/Users/egonat/repos/status-im/nim-codex/codex/logutils.nim imports /Users/egonat/repos/status-im/nim-codex/codex/codex.nim
/Users/egonat/repos/status-im/nim-codex/codex/codex.nim imports /Users/egonat/repos/status-im/nim-codex/codex/logutils.nim
/Users/egonat/repos/status-im/nim-codex/codex/logutils.nim imports /Users/egonat/repos/status-im/nim-codex/codex/logutils.nim
/Users/egonat/repos/status-im/nim-codex/codex/logutils.nim imports /Users/egonat/repos/status-im/nim-codex/codex/logutils.nim
/Users/egonat/repos/status-im/nim-codex/codex/logutils.nim imports /Users/egonat/repos/status-im/nim-codex/codex/codex.nim
/Users/egonat/repos/status-im/nim-codex/codex/codex.nim imports /Users/egonat/repos/status-im/nim-codex/codex/node.nim
/Users/egonat/repos/status-im/nim-codex/codex/node.nim imports /Users/egonat/repos/status-im/nim-codex/codex/chunker.nim
/Users/egonat/repos/status-im/nim-codex/codex/chunker.nim imports /Users/egonat/repos/status-im/nim-codex/codex/blocktype.nim
/Users/egonat/repos/status-im/nim-codex/codex/blocktype.nim imports /Users/egonat/repos/status-im/nim-codex/codex/codextypes.nim
/Users/egonat/repos/status-im/nim-codex/codex/codextypes.nim imports /Users/egonat/repos/status-im/nim-codex/codex/units.nim
/Users/egonat/repos/status-im/nim-codex/codex/units.nim imports /Users/egonat/repos/status-im/nim-codex/codex/logutils.nim

```
`logutils` does not import `codex` which seems to be the root of the circular dependency.

Reply:
this looks bad overall, have we figured out the root cause?

Reply:
By removing the `{.dirty.}` annotation and injecting the needed `it` vars, the issue was resolved. `questionable/results` is also now exported, so its possible this was the cause of the undeclared identifier.